source/Ship.cpp

```diff
@@ -1995,6 +2010,52 @@ bool Ship::FireAntiMissile(const Projectile &projectile, vector<Visual> &visuals
 
 
 
+// Fire tractor beams at the given flotsam. Returns a Point representing the net
+// pull on the flotsam from this ship's tractor beams.
+Point Ship::FireTractorBeam(const Flotsam &flotsam, vector<Visual> &visuals)
+{
+	Point pullVector;
+	if(flotsam.Position().Distance(position) > tractorBeamRange)
+		return pullVector;
+	if(CannotAct())
+		return pullVector;
+	// Don't waste energy on flotsams that you can't pick up.
+	if(!CanPickUp(flotsam))
+		return pullVector;
+	if(IsYours())
+	{
+		const auto flotsamSetting = Preferences::GetFlotsamCollection();
+		if(flotsamSetting == Preferences::FlotsamCollection::OFF)
+			return pullVector;
+		if(!GetParent() && flotsamSetting == Preferences::FlotsamCollection::ESCORT)
+			return pullVector;
+		if(flotsamSetting == Preferences::FlotsamCollection::FLAGSHIP)
+			return pullVector;
+	}
+
+	double jamChance = CalculateJamChance(Energy(), scrambling);
+
+	bool opportunisticEscorts = !Preferences::Has("Turrets focus fire");
+	const vector<Hardpoint> &hardpoints = armament.Get();
+	for(unsigned i = 0; i < hardpoints.size(); ++i)
+	{
+		const Weapon *weapon = hardpoints[i].GetOutfit();
+		if(weapon && CanFire(weapon))
+			if(armament.FireTractorBeam(i, *this, flotsam, visuals, Random::Real() < jamChance))
+			{
+				Point hardpointPos = Position() + Zoom() * Facing().Rotate(hardpoints[i].GetPoint());
+				// Heavier flotsam are harder to pull.
+				pullVector += (hardpointPos - flotsam.Position()).Unit() * weapon->TractorBeam() / flotsam.Mass();
+				// If this ship is opportunistic, then only fire one tractor beam at each flostam.
+				if(personality.IsOpportunistic() || (isYours && opportunisticEscorts))
+					break;
+			}
+	}
+	return pullVector;
```

Comment:
Thoughts?
```suggestion
	// If any tractor beams successfully fired on this flotsam, also drag the flotsam with
	// the ship's velocity. This makes tractor beams feel much more capable of dragging
	// flotsams to the ship. Otherwise, a ship could be drifting away from a flotsam
	// at the same speed that the tractor beam is pulling the pulling the flotsam
	// toward the ship, which looks awkward. This also allows faster ships to "steal"
	// flotsams from slower ones without the need of a stronger tractor beam. Although
	// this does mean that if you fly toward a flotsam that is in your tractor beam,
	// you're pushing the flotsam away from your ship (although it will still slowly
	// come toward you from the pull of the tractor beam itself).
	if(pullVector)
		pullVector += velocity;
	return pullVector;
```
This would likely require AI changes so that ships with a flotsam in their tractor beam aren't fruitlessly chasing it. This also looks a little odd if multiple ships are pulling on the same flotsam, but I think it feels much better for a single ship pulling on a flotsam.

Reply:
would be a cool quality of life thing, dunno how much AI changes would be necessary, probably not too much?

Reply:
AI::DoHarvesting would need to have some understanding of which flotsams a ship is targeting with its tractor beams, and then not target those to chase. I think the cleanest method would be for Ship to have a set of flotsams that it's using its tractor beams on (populated during FireTractorBeams), then AI can check if a flotsam is in that set and skip it if it is. This set would need to be cleared every frame, so we could probably clear it Engine::MoveShip.

Reply:
Trying to implement this, but turns out just adding the velocity of each ship firing a tractor beam on a flotsam can cause the ships to kick the flotsam away from them if they're all moving in the same direction toward the flotsam. Need a more elegant way of breaking physics to make the tractor beams worthwhile.